Allow settings_files to be a list of pathlib.Paths

I just came across this little project, looks really nice!
What I don't get yet is the need for the `settings_files` to be strings, rather than `pathlib.Path`:

https://github.com/daviskirk/climatecontrol/blob/fc84af82205ac7eb728406f542efcb8237293ea1/climatecontrol/core.py#L193

For example I would like to use it like this:

```python
from pathlib import Path
from climatecontrol import Climate

climate = Climate(settings_files=[Path.home() / ".config" / "myapp" / "myapp.json"])
```

but what I get from `mypy` is:

```
$ poetry run mypy --namespace-packages -p myapp
myapp/cli.py:9: error: List item 0 has incompatible type "Path"; expected "str"
Found 1 error in 1 file (checked 3 source files)
```

--- a/climatecontrol/core.py
+++ b/climatecontrol/core.py
@@ -140,7 +140,7 @@ class SettingsItem(ObjectProxy):
 
     @classmethod
     def _self_is_mutable(cls, value: Any) -> bool:
-        return isinstance(value, MutableMapping) or isinstance(value, MutableSequence)
+        return isinstance(value, (MutableMapping, MutableSequence))
 
 
 class Climate:
@@ -177,6 +177,7 @@ class Climate:
 
     """
 
+    settings_files: List[Union[str, Path]]
     _combined_fragment: Fragment
     _updates: List
     _fragments: List[Fragment]
@@ -190,14 +191,17 @@ class Climate:
 
     def __init__(
         self,
-        settings_files: Sequence[str] = (),
+        settings_files: Union[str, Path, Sequence[Union[str, Path]]] = (),
         parser: Optional[Callable[[Mapping], Mapping]] = None,
         **env_parser_kwargs: Any,
     ) -> None:
         """Initialize settings object."""
         self.env_parser = EnvParser(**(env_parser_kwargs or {}))
         self.parser = parser
-        self.settings_files = settings_files  # type: ignore
+        if isinstance(settings_files, (str, Path)):
+            self.settings_files = [settings_files]
+        else:
+            self.settings_files = list(settings_files)
         self._updates = []
         self._fragments = []
         self._initialized = False
@@ -209,15 +213,6 @@ class Climate:
     def __repr__(self) -> str:
         return self.__class__.__qualname__ + "[\n{}\n]".format(pformat(self._data))
 
-    def __str__(self) -> str:
-        return self.__class__.__qualname__ + "[\n{}\n]".format(pformat(self._data))
-
-    @property
-    def settings(self) -> Any:
-        """Return a settings item proxy for easy access to settings hierarchy."""
-        self.ensure_initialized()
-        return SettingsItem(self._data, self, FragmentPath())
-
     @property
     def parser(self) -> Optional[Callable[[Mapping], Mapping]]:
         """Return settings parser function."""
@@ -226,22 +221,13 @@ class Climate:
     @parser.setter
     def parser(self, value: Optional[Callable[[Mapping], Mapping]]) -> None:
         """Set the settings parser function."""
-        if value and not callable(value):
-            raise TypeError("If given, ``parser`` must be a callable")
         self._parse = value
 
     @property
-    def settings_files(self) -> List[str]:
-        """Return current settings files."""
-        return self._settings_files
-
-    @settings_files.setter
-    def settings_files(self, files: Union[str, Iterable[str]]) -> None:
-        """Set settings files to use when parsing settings."""
-        files = files or ()
-        if isinstance(files, str):
-            files = (files,)
-        self._settings_files = list(files)
+    def settings(self) -> Any:
+        """Return a settings item proxy for easy access to settings hierarchy."""
+        self.ensure_initialized()
+        return SettingsItem(self._data, self, FragmentPath())
 
     @property
     def inferred_settings_files(self) -> List[Path]:
@@ -493,15 +479,15 @@ class Climate:
 
         """
         archived_data = deepcopy(self._data.__wrapped__)
-        archived_settings = {}
-        for k in [
-            "settings_files",
-            "_updates",
-            "_fragments",
-            "_combined_fragment",
-        ]:
-            archived_settings[k] = deepcopy(getattr(self, k))
-
+        archived_settings = {
+            k: deepcopy(getattr(self, k))
+            for k in [
+                "settings_files",
+                "_updates",
+                "_fragments",
+                "_combined_fragment",
+            ]
+        }
         yield self
 
         # reinstate all saved data after context block is finished

--- a/climatecontrol/ext/dataclasses.py
+++ b/climatecontrol/ext/dataclasses.py
@@ -74,5 +74,5 @@ class Climate(BaseClimate, Generic[T]):
     def parse(self, data: Mapping) -> T:
         """Parse data into the provided dataclass."""
         data = super().parse(data)
-        obj: T = dacite.from_dict(self.dataclass_cls, data)
+        obj: T = dacite.from_dict(self.dataclass_cls, {k: v for k, v in data.items()})
         return obj

--- a/mypy.ini
+++ b/mypy.ini
@@ -1,5 +1,4 @@
 [mypy]
 ignore_missing_imports=True
-follow_imports=skip
 disallow_untyped_defs=False
 check_untyped_defs=True

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "climatecontrol"
-version = "0.9.2"
+version = "0.10.0"
 description = "Python library for loading app configurations from files and/or namespaced environment variables."
 authors = ["Davis Kirkendall <davis.e.kirkendall@gmail.com>"]
 license = "MIT"

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -195,9 +195,3 @@ def mock_env_settings_file(mock_os_environ, mock_settings_file):
     """Set the settings file env variable to a temporary settings file."""
     os.environ["TEST_STUFF_SETTINGS_FILE"] = mock_settings_file[0]
     return mock_settings_file
-
-
-@pytest.fixture
-def mock_env_parser(mocker):
-    """Mock out environment variable parser."""
-    return mocker.patch("climatecontrol.core.EnvParser")

--- a/tests/test_env_parser.py
+++ b/tests/test_env_parser.py
@@ -145,6 +145,6 @@ def test_env_parser_iter_load(monkeypatch, environ, expected_kw):
     else:  # pragma: nocover
         # python 3.5 doesn't order dicts so we can't test the exact order
         def to_set(fragments):
-            return set(str(f) for f in fragments)
+            return {str(f) for f in fragments}
 
         assert to_set(result) == to_set(expected)

--- a/tests/test_settings.py
+++ b/tests/test_settings.py
@@ -20,6 +20,7 @@ def test_settings_empty(mock_empty_os_environ):
     climate = core.Climate()
     assert isinstance(climate.settings, Mapping)
     assert dict(climate.settings) == {}
+    assert repr(climate)  # check that __repr__ works
     assert str(climate)  # check that __repr__ works
     assert str(climate.settings)  # check that __repr__ works
     assert len(climate.settings) == 0  # length of settings map
@@ -542,22 +543,6 @@ def mock_parser_fcn(s):
     """Return input instead of doing some complex parsing."""
 
 
-@pytest.mark.parametrize(
-    "attr, value, expected",
-    [
-        ("settings_files", "this.toml", ["this.toml"]),
-        ("settings_files", ("this.toml", "that.toml"), ["this.toml", "that.toml"]),
-        ("parser", mock_parser_fcn, mock_parser_fcn),
-    ],
-)
-def test_assign(mock_empty_os_environ, mock_env_parser, attr, value, expected):
-    """Test that assigning attributes on settings object works."""
-    s = core.Climate(prefix="this", settings_file_suffix="suffix", parser=None)
-    assert s.settings_files == []
-    setattr(s, attr, value)
-    assert getattr(s, attr) == expected
-
-
 @pytest.mark.parametrize("update", [False, True])
 @pytest.mark.parametrize("envvar", [False, True])
 @pytest.mark.parametrize("clear", [False, True])
@@ -752,7 +737,7 @@ def test_setup_logging(monkeypatch, update, mock_empty_os_environ):
 def test_update_log(caplog, mock_empty_os_environ, mock_settings_file):
     """Test writing out an example configuration file."""
     settings_file_path, expected = mock_settings_file
-    climate = core.Climate(prefix="TEST_STUFF", settings_files=settings_file_path)
+    climate = core.Climate(prefix="TEST_STUFF", settings_files=(settings_file_path,))
     assert climate.update_log == "", "before updating, the update log should be empty"
     climate.update({"a": core.REMOVED, "b": 2})
     lines = climate.update_log.split("\n")

